File: 13.gRPC-and-Inter-Service-Communication-for-Microservices/code/typescript/client_streaming_4.ts
// proto:  rpc UploadEvents(stream Event) returns (UploadSummary);
import * as grpc from '@grpc/grpc-js';

// ===== SERVER: on('data') in a loop, then callback() once at the end =====
class UserServer {
    uploadEvents(call: grpc.ServerReadableStream<any, any>, callback: grpc.sendUnaryData<any>) {
        let count = 0;
        
        call.on('data', (ev: any) => {
            store(ev);
            count++;
        });
        
        call.on('end', () => {
            // client finished sending, now reply once
            const summary = new UploadSummary();
            summary.setReceived(count);
            callback(null, summary);
        });
        
        call.on('error', (err) => {
            callback(err, null);
        });
    }
}

// ===== CLIENT: write() many, then end() to await the single reply =====
function uploadEvents(client: any, events: any[]) {
    const deadline = new Date();
    deadline.setSeconds(deadline.getSeconds() + 30);
    
    const stream = client.uploadEvents({ deadline }, (err: any, summary: any) => {
        // close our side, await the summary
        if (err) {
            console.error(err);
            return;
        }
        console.log(`server received ${summary.getReceived()} events`);
    });
    
    for (const ev of events) {
        stream.write(ev); // push each one up
    }
    
    stream.end(); // signal we're done sending
}

// Dummy definitions
class UploadSummary { setReceived(c: number){} getReceived(){ return 0; } }
function store(e: any){}

File: 13.gRPC-and-Inter-Service-Communication-for-Microservices/code/rust/client_streaming_4.rs
// proto:  rpc UploadEvents(stream Event) returns (UploadSummary);

use tonic::{Request, Response, Status};

// Dummy types
pub struct Event {}
pub struct UploadSummary { pub received: i32 }
pub struct MyServer;
fn store(e: Event) {}

// ===== SERVER: next() in a loop, then Ok(Response) once at the end =====
#[tonic::async_trait]
impl MyServer {
    async fn upload_events(
        &self,
        request: Request<tonic::Streaming<Event>>,
    ) -> Result<Response<UploadSummary>, Status> {
        let mut stream = request.into_inner();
        let mut count = 0;
        
        while let Some(ev) = stream.message().await? {
            // client finished sending when stream.message() returns None, now reply once
            store(ev);
            count += 1;
        }
        
        Ok(Response::new(UploadSummary { received: count }))
    }
}

// ===== CLIENT: stream of requests, await the single reply =====
async fn upload_events(mut client: AnyServiceClient, events: Vec<Event>) {
    // create a stream of requests
    let iter = events.into_iter().map(Ok);
    let request_stream = futures::stream::iter(iter);
    
    let mut req = Request::new(request_stream);
    req.set_timeout(std::time::Duration::from_secs(30));

    // close our side, await the summary
    let summary = client.upload_events(req).await.unwrap().into_inner();
    
    println!("server received {} events", summary.received);
}
struct AnyServiceClient;
impl AnyServiceClient { async fn upload_events<T>(&mut self, req: Request<T>) -> Result<Response<UploadSummary>, Status> { unimplemented!() } }
